✨using shouldComponentUpdate from React 16.3.x

umbertoghio opened the pull request:
Hello, 
as I am getting a better understanding of how internals of ReactN works I switched from the UNSAFE_componentWillUpdate to shouldComponentUpdate only if using React 16.3.x onwards, so the behavior for older Reacts is unchanged.

The only difference in behavior between componentWillUpdate and shouldComponentUpdate is that the first doesn't return anything while the second must return true if you want to simulate a function not been defined un user's class.

This should ensure compatibility with React 17.x onwards without using UNSAFE's methods.

Still a NOOB on TypeScript, apologizes if I used improper syntax, but all checks pass.

CharlesStover commented on src/utils/should-component-update.ts:
```diff
@@ -0,0 +1,57 @@
+    that.shouldComponentUpdate = (...args: [ P, S, any ]): boolean => {
+      ReactNCShouldComponentUpdate(that);
+      return proto.shouldComponentUpdate.bind(that)(...args); // Returns outcome of shouldComponentUpdate method
+    };
```
It should `return true` here, similar to how [`componentWillUpdate` returns](https://github.com/CharlesStover/reactn/blob/master/src/utils/component-will-update.ts#L61).

This check is used where you call `if (!shouldComponentUpdatePrototype) { that.shouldComponentUpdate = () => {}` in another file in this PR.

This file reads, "If this component DOES have a shouldComponentUpdate method on its prototype, set a custom shouldComponentUpdate method on its instance that clears the global subscriptions and then calls the prototype method."

The other file reads, "If this component DOES NOT have a shouldComponentUpdate method on its prototype (and as a result, does not clear subscriptions on its instance), then create a shouldComponentUpdate method on the instance (and don't call anything on the prototype)."

CharlesStover commented:
Awesome, thank you. I left one comment where you are missing a return statement. Other than that, it looks good. :)

umbertoghio commented:
I missed it indeed, I got confused as the should function requires a return too :-)

checking the tests output more carefully, however, I noticed this:

`Warning: TestMount has a method called shouldComponentUpdate(). shouldComponentUpdate should not be used when extending React.PureComponent. Please extend React.Component if shouldComponentUpdate is used.`

See https://ozmoroz.com/2018/09/what-is-purecomponent/

This solution is still not clean as it may impact PureComponents....😓

What about getDerivedStateFromProps method ?
I would give it a go but I get 
`Property 'getDerivedStateFromProps' does not exist on type 'Component<P, S, any>'.`  
Typescript error and I can't figure it out....

What do you suggest? 
Maybe just stick with UNSAFE_componentwillmount ?

CharlesStover commented:
`getDerivedStateFromProps` is `static`, so it exists on the class but not the instance of the class, which unfortunately makes it impossible for the _instance_ to unsubscribe from the global state. When `getDerivedStateFromProps` is called, it does not know which instance of a component for which it is being called.

That's a very good catch regarding pure components. I'm not sure the best solution in that case. I want to get off `unsafe` methods where possible. If you want a quick solution, it may be possible to bind `shouldComponentUpdate` for `Component` and `UNSAFE_componentWillUpdate` for `PureComponent` for the time being. That will at least get `Component` off of the unsafe method, and we can address PureComponent later, if a solution can be discovered.

umbertoghio commented:
Hi, thank you for the explanation it's clear 😄

I have implemented your suggestion: 
if old version -> componentWillUpdate
If new version && regular -> shouldComponentUpdate
if new version && Pure => UNSAFE_componentWillUpdate

But there's a Catch:
during dev I modified the test to check for Pure and regular components, the logic was seemingly OK for Pure but I later found out that shouldCompoenntUpdate was never firing on regular components. Not sure if was something about the tests or it also happens for real components so to be on the safe side I commented out a bit of logic to obtain the following outcome for Decorated components:
if old version -> componentWillUpdate
If new version && regular -> UNSAFE_componentWillUpdate
if new version && Pure => UNSAFE_componentWillUpdate

CharlesStover commented on src/decorator.ts:
```diff
@@ -58,26 +63,34 @@ export default function ReactN<
-    public UNSAFE_componentWillUpdate(...args: [ P, S, any ]): void {
-      const [ rVerMaj, rVerMin ] = version.split('.').map((v): number => parseInt(v));
-      if (rVerMaj > 16 || (rVerMaj === 16 && rVerMin >= 3)) {
+    public componentWillUpdate(...args: [ P, S, any ]): void {
```
I think this will console log ("Use UNSAFE instead"), since you are defining this method. This method should only _exist_ if `isUsingOldReact`.

CharlesStover commented on src/decorator.ts:
```diff
@@ -21,6 +22,10 @@ const isComponentDidMount = false;
 
+const [ rVerMaj, rVerMin ] = version.split('.').map((v): number => parseInt(v));
+const isUsingOldReact = rVerMaj < 16 || (rVerMaj === 16 && rVerMin < 3);
+const isPureComponent = (name: string): boolean => name === 'DecoratedCwuPrototype-ReactN'
```
Where is `DecoratedCwuPrototype` coming from?

CharlesStover commented on src/decorator.ts:
```diff
@@ -110,5 +123,10 @@ export default function ReactN<
   */
 
+  if (isPureComponent(DecoratedReactNComponent.displayName)) {
+    // Must remove shouldComponentUpdate from prototype for pure components or it will get warning and side effects
```
Is this still needed, since we aren't adding `shouldComponentUpdate` to decorated components? You commented out the method above. At this point, we are deleting a method added by a third party (React itself or the developer themselves) and not a method we created.

CharlesStover commented:
>during dev I modified the test to check for Pure and regular components, the logic was seemingly OK for Pure but I later found out that shouldCompoenntUpdate was never firing on regular components. Not sure if was something about the tests or it also happens for real components

There's no reason the test should be different than real components. The tests are ran for _many_ different versions of React, so pay attention to the version that failed -- it could be that `shouldComponentUpdate` does not exist in that version.

umbertoghio commented:
Hi, I did some more tests and I can confirm the shouldComponentUpdate override for Decorated components is not firing on the latter React versions (16.3, 16.8), I can't explain why...
So to bring it to a consistent state I reverted to using UNSAFE_componentWillMount with some cleanup. Hope the steps done so far was of some help 😄
